### ISSUE TITLE
[otbn, sw] computation of q3 in barrett384.s

### ISSUE THREAD

**Issue by yizhou7** [lowRISC/opentitan#4549]
Starting from [here](https://github.com/lowRISC/opentitan/blob/4a8eea22f7e4dbb6c986126970cf37e6903871c8/sw/otbn/code-snippets/barrett384.s#L167), in the first step:

```
  /* Add q1. This is unconditional since MSb of u is always 1.
     This cannot overflow due to leading zeros.
     q2''' = q2'' + q1 = [w20, w19] = [w20, w19] + [w8, w9] */
  bn.add w19, w19, w8
  bn.addc w20, w20, w9
```

`q1` is a 385 bit number from the previous step, and `q2''` is a 384 bit number, so `q2'''` could use 386 bits.

In the next step:

```
  /* Conditionally add u (without leading 1) in case of MSb of x being set.
     This is the "real" q2 but shifted by 384 bits to the right. This cannot
     overflow due to leading zeros
     q2'''' = x[767]?q2'''+u[383:0]:q2'''
            = [w20, w19] + [w25, w24] = q2 >> 384 */
  bn.add w19, w19, w24
  bn.addc w20, w20, w25
  ```
Another 384 bit number is "conditionally" added to `q2'''`. Therefore `q2''''` should still fit into 386 bits.

In the next step:
```
/* finally this gives q3 by shifting the remain bit to the right
	q3 = q2 >> 385 = q2'''' >> 1 = [w9, w8] = [w20, w19] >> 1 */
bn.rshi w9, w20, w31 >> 1
bn.rshi w8, w19, w20 >> 1
```
`q2'''' ` is right shifted to compute `q3`, so `q3 = [w9, w8]` should fit into 385 bits. 

In the next step:
```
/* Compute r2 = q3 * m % 2^385.
	=> max. length r2: 385 bit
	q3*m = [w18, w17, w16] = [w9, w8] * [w13,w12] */
bn.mov w10, w12
bn.mov w11, w13
jal x1, mul384
```
The multiplication works for 384 bits numbers. However `q3` is 385 bits. Let `b` be the most significant bit of `q3` and `q3'` be the rest of the 384 bits. So `q3 = b * 2^384 + q3'`. It seems like the multiplication computes `q3' * m`, instead of  `q3 * m`.  We note that `r` only cares about `q3 * m % 2^385`, so we check:
```
q3 * m % 2^385 
= (b * 2^384 + q3') * m % 2^385
= (b * 2^384 * m + q3' * m) % 2^385
```
Hence, it seems that when `b = 1`, unless `m` (which is the modulo) is even, the remainders are not equal. Please correct me if I made a mistake somewhere.  Thank you, @felixmiller.

> estimate 4
> remaining 2021-01-25 0

**Comment by felixmiller** [comment]
Thanks @yizhou7 for the the detailed analysis. You are fully right, if `q3` takes up 385 bit, using the 384 bit multiplication kernel would be wrong. I wasn't able to find a counter example though and had a closer look on the original 256 bit implementation we derived this code from and which also only uses a 256 bit (not 257 bit in this case) multiplication kernel.

I think the justification is another implicit assumption that we don't point out in the comments clearly at the moment (what we should). That is, that both input operands `a` and `b` must never exceed the modulus `m`.

In this case I believe that `q3` can never consume more than 384 bit.

Here is my reasoning. Would be great if you could check for correctness and maybe also consider in you proofing tools.
`q3` is only 384 bit wide (or less) if `q3 < 2^384`
with `q3 = floor(floor(x/2^383)*floor(2^768/m)*1/2^385)`

The last term together with the outer `floor` function is a simple 385 bit right shift. Hence in order for `q3` being 385 bit,
`q2` which is `q2 = floor(x/2^383)*floor(2^768/m)` had to be 770 bit, and the inequality that must hold is now `q2<2^769`.

A larger `x` results in a larger `q2`. Since `x=a*b`, for a maximized `q2`, the operands `a` and `b` can be set to `a=b=m-1` and we can write `q2 = floor((m-1)^2/2^383)*floor(2^768/m) < 2^769`.

Both floor function can only result in a smaller q2, so we can remove them. or in other words, if we can show that `q2` is never 770 bit wide without the floor functions, it also won't be with them.

```
q2 = (m-1)^2/2^383 * 2^768/m = (m-1)^2/m*2^385 < 2^769
<=>  (m-1)^2/m < 2^384
```
The subtraction of 1 also always reduces the left side, so if we can show that `m^2/m = m < 2^384` holds, the inequality above also holds.

The maximum value of `m` (per definition, input width) is `m = 2^384-1`. So if the above reasoning is correct, we can conclude that `q3` never is wider than 384 bit.

This is different in a pure reduction algorithm, where there is no such bound for the input value other than that it can't be wider than 768 bit. The HAC 14.42 algorithm is actually a pure reduction algorithm and does not include the computation of the product. This is something that we should more explicit in the comments. So I will rework the header comment and also include the upper bounds on the operands.

**Comment by felixmiller** [comment]
Additional comments w.r.t. to the assumptions on the operands are now added in [this](https://github.com/lowRISC/opentitan/pull/4957) PR next to other fixes as described in #4548.

**Comment by imphil** [comment]
Fixed by https://github.com/lowRISC/opentitan/pull/4957
